components/brave_wallet/browser/zcash/discover_next_unused_zcash_address_task.cc

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_wallet/browser/zcash/discover_next_unused_zcash_address_task.h"
+
+#include "brave/components/brave_wallet/common/common_utils.h"
+#include "brave/components/brave_wallet/common/zcash_utils.h"
+#include "components/grit/brave_components_strings.h"
+#include "ui/base/l10n/l10n_util.h"
+
+namespace brave_wallet {
+
+DiscoverNextUnusedZCashAddressTask::DiscoverNextUnusedZCashAddressTask(
+    base::WeakPtr<ZCashWalletService> zcash_wallet_service,
+    mojom::AccountIdPtr account_id,
+    mojom::ZCashAddressPtr start_address,
+    ZCashWalletService::DiscoverNextUnusedAddressCallback callback)
+    : zcash_wallet_service_(std::move(zcash_wallet_service)),
+      account_id_(std::move(account_id)),
+      start_address_(std::move(start_address)),
+      callback_(std::move(callback)) {}
+
+DiscoverNextUnusedZCashAddressTask::~DiscoverNextUnusedZCashAddressTask() =
+    default;
+
+void DiscoverNextUnusedZCashAddressTask::Iterate() {
+  base::SequencedTaskRunner::GetCurrentDefault()->PostTask(
+      FROM_HERE,
+      base::BindOnce(&DiscoverNextUnusedZCashAddressTask::WorkOnTask, this));
```

Comment:
@cypt4 should `this` be replaced with `weak_ptr_factory_.GetWeakPtr()` since the ZCashWalletService could be called after it's freed?

Reply:
https://chromium.googlesource.com/chromium/src/+/main/docs/callback.md#binding-a-class-method-with-weak-pointers is the docs I was looking at when reviewing this

Reply:
not exactly. If `BindOnce/BindRepeating` was compiled without errors with `this` as the first bound argument, then only two of these are possible:
1. the function is not a class member or is a static function that just receives a raw pointer as the first argument. Ex:
```
void MyFunc(SomeClass*);

static void SomeClass::StaticClassFunc(SomeClass*);

void SomeClass::DoStuff() {
  // this is "fine" and will compile, but smells bad and most likely will break, we should avoid that.
  base::BindOnce(&MyFunc, this);

  // this is "fine" and will compile, but also is very weird and will most likely break at some point.
  base::BindOnce(&SomeClass::StaticClassFunc, this);
}
```
2. the function is a class member, and the class is RefCounted.

in this exact case we see a pointer to a class member function. This function is a non-static function, the class is RefCounted, meaning `Bind` is safe as it will keep the reference to the class inside the callback.

Reply:
Ok, from a sec perspective it sounds like not an issue then. If you have some suggestions as to how to make this more robust feel free to add them, but I don't think this requires blocking over based on your response @goodov